### PR TITLE
Feature/dwr 1675 UI for IAB cards

### PR DESCRIPTION
**Pull request by dsnyder-fairway**
a typical group with URL

screen shot 2018-04-18 at 4 03 26 pm

how to access this view

screen shot 2018-04-18 at 4 03 57 pm

screen shot 2018-04-18 at 4 04 10 pm

viewing math 

screen shot 2018-04-18 at 4 04 40 pm

selecting some cards

screen shot 2018-04-18 at 4 04 48 pm

### CODE REVIEW

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/filter/DefaultExamFilterOptionService.java [hunk 1]
```diff
@@ -17,17 +18,20 @@
...
+    private final SubjectService subjectService;
 
     @Autowired
     public DefaultExamFilterOptionService(
             @NotNull final SchoolYearService schoolYearService,
             @NotNull final EthnicityService ethnicityService,
             @NotNull final GenderService genderService,
-            @NotNull final ElasService elasService) {
+            @NotNull final ElasService elasService,
+            @NotNull final SubjectService subjectService) {
```
Comment (dsnyder-fairway) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/filter/DefaultExamFilterOptionService.java]: grab list of subjects

**Comment by dsnyder-fairway** [comment]
note -- there will be continue work on this

File: webapp/src/main/webapp/src/app/group-cards/group-card.service.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import { Group } from '../groups/group';
+import { AssessmentExamMapper } from '../assessments/assessment-exam.mapper';
+import { MeasuredAssessment } from './measured-assessment';
+
+const ServiceRoute = ReportingServiceRoute;
+
+@Injectable()
+export class GroupCardService {
+  group: Group;
+  schoolYear: number;
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/group-cards/group-card.service.ts]: think you forgot to remove these

File: webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    this._currentSubject = value;
+    if (this._currentSubject !== 'ALL') {
+      this.measuredAssessments = this.measuredAssessments.filter(
+        measuredAssessment => measuredAssessment.assessment.subject === this._currentSubject);
+    }
+  }
+
+  private updateFilterOptions(): void {
+    this.filterOptions.subjects.push('ALL');
+    this.filterOptions.subjects.sort((a, b) => a.localeCompare(b));
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts]: we should make a dedicated service that does this bit of logic - also we should not alpha sort - we should use the natural order of the backend (Math, ELA)

File: webapp/src/main/webapp/src/app/groups/group.ts [hunk 4]
```diff
@@ -1,7 +1,8 @@
...
-  schoolName: string;
-  schoolId: number;
-  subjectCode: string;
+export interface Group {
+  readonly id: number;
+  readonly name: string;
+  readonly schoolName: string;
+  readonly schoolId: number;
+  readonly subjectCode: string;
+  readonly totalStudentCount?: number;
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/groups/group.ts]: i'd pick either `totalStudents` or `studentCount` i think total and count are a bit redundant

File: webapp/src/main/webapp/src/app/group-cards/measured-assessment.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { Assessment } from '../assessments/model/assessment.model';
+
+export interface MeasuredAssessment {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/group-cards/measured-assessment.ts]: 👍

File: webapp/src/main/webapp/src/app/group-cards/iab-card.component.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+  percentFill(index: number): number {
+    return Math.round((this.measuredAssessment.studentCountByPerformanceLevel[ index ] / this.measuredAssessment.studentsTested) * 100);
+  }
+
+  selectCard(): void {
+    this.selected = !this.selected;
+    this.selectedIab.emit(<IabEvent>{
+      id: this.measuredAssessment.assessment.id,
+      selected: this.selected
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/group-cards/iab-card.component.ts]: i'd be tempted to call these "assessment cards" and "assessment card events" also i think it might make more sense to have a card model which holds 1. the assessment info and 2. whether or not it is selected or not. then you can just dispatch the card object when clicked

File: webapp/src/main/webapp/src/app/group-cards/iab-card.component.html [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+      <div class="bar-container">
+        <div class="bar blue-dark" *ngIf="percentFill(0) > 0" [attr.data-width]="percentFill(0)">
+          <span class="percentage">{{ 'common.value-display-type-format.Percent' | translate: { value: percentFill(0)} }}</span>
+          <span class="bar-label sr-only">{{ 'common.assessment-type.iab.performance-level.1.name' }}</span>
+        </div>
+        <div class="bar blue-dark aqua" *ngIf="percentFill(1) > 0" [attr.data-width]="percentFill(1)">
+          <span class="percentage">{{ 'common.value-display-type-format.Percent' | translate: { value: percentFill(1)} }}</span>
+          <span class="bar-label sr-only">{{ 'common.assessment-type.iab.performance-level.2.name' }}</span>
+        </div>
+        <div class="bar aqua" *ngIf="percentFill(2) > 0" [attr.data-width]="percentFill(2)">
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/group-cards/iab-card.component.html]: i would pre-compute these kind of things and shove them into a view model so they don't have to be computed redundantly at render time.

**Comment by jtreuting** [comment]
Quick note but will review more tonight.  The CSS for the selected cards is missing the blue circle and icon in the upper right that the code pen had.

File: webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  addIab(event: IabEvent) {
+    if (event.selected) {
+      this.selectedIabs.push(event.id);
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts]: would change this to `selectedCards` and just hold the card in its entirety no?

File: webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+      this.disableView = false;
+    } else {
+      this.selectedIabs = this.selectedIabs.filter(id => id !== event.id);
+      if (this.selectedIabs.length === 0) {
+        this.disableView = true;
+      }
+    }
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts]: i would rewrite this whole method as something like:
```typescript

onCardSelectionChange() {
  this.selectedCards = this.cards.filter(card => card.selected);
  this.disableView = this.selectedCards.length === 0;
}
```

File: webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts (hunk 8 above)
Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts]: my understanding was that this would be used to go to the old groups pages with the selected assessments displaying. That's why I am just passing the id for now

File: webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  get currentGroup(): Group {
+    return this._currentGroup;
+  }
+
+  set currentGroup(value: Group) {
+    this._currentGroup = value;
+    if (this._currentGroup) {
+      this.groupCardService.group = this._currentGroup;
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts]: why is this necessary? can't we just pass the current group into the service method call?

File: webapp/src/main/webapp/src/app/group-cards/group-cards.component.html [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+                [(ngModel)]="currentSubject"
+                class="form-control"
+                (change)="updateRoute()">
+          <option *ngFor="let subject of filterOptions.subjects"
+                  [ngValue]="subject">{{'common.subject.' + subject + '.short-name' | translate}}
+          </option>
+        </select>
+      </div>
+      <div class="col-md-6">
+        <a href="javascript:void()" [ngClass]="{'disabled': disableView}" class="col-btn btn btn-primary">{{ 'iab-card.view-assessments' | translate }}<i class="ml-xs fa fa-angle-right"></i></a>
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/group-cards/group-cards.component.html]: i would flip this var to be `cardViewEnabled`

**Comment by dsnyder-fairway** [comment]
@jtreuting looks like the latest codepen has this in there. I'll have to grab the latest

**Comment by esotrope** [comment]
should the cards have a white bg?

File: webapp/src/main/webapp/src/app/app.routes.ts [hunk 12]
```diff
@@ -216,6 +217,22 @@ export const routes: Routes = [
           studentTestHistoryChildRoute
         ]
       },
+      {
+        path: 'group-cards/:groupId',
```
Comment (jtreuting) [inline webapp/src/main/webapp/src/app/app.routes.ts]: Should we maybe can this the group-dashboard?

File: webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+export class GroupCardsComponent implements OnInit {
+
+  measuredAssessments: MeasuredAssessment[] = [];
+  group: Group;
+  groups: Group[];
+  _currentGroup: Group;
+  filterOptions: ExamFilterOptions = new ExamFilterOptions();
+  currentSchoolYear: number;
+  _currentSubject: string;
+  disableView = true;
```
Comment (jtreuting) [inline webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts]: Do we have a standard for the _ vs non names?  I'm having a hard time seeing when one is used versus the other

Comment (jtreuting) [reply webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts]: Send like the underscore should be private?

Comment (esotrope) [reply webapp/src/main/webapp/src/app/group-cards/group-cards.component.ts]: i think we've been only using _ on private members that need getters/setters. might be more straight forward if we always use it for private members though

**Comment by dsnyder-fairway** [comment]
going to close this since I renamed a bunch of files and did a bunch of improvements. I'll issue a separate PR once UT have been completed
